### PR TITLE
Fix #104794: Fixed sret bugs with `clang::musttail` on 32 bit targets

### LINKED ISSUE #104794: For tail calls, clang generates extra alloca for sret parameter on 32 bit targets
While running a fuzzer to verify #102896, I kept getting errors for any cases using `double _Complex` or `_BitInt(n>64)` as return types for `[[clang::musttail]]` functions on 32 bit targets. These cases would compile such that new memory is `alloca`'d  for the sret parameter, rather than passing through the same sret ptr from the caller. For example

```cpp
double _Complex F11(signed short P0);

double _Complex F12(signed short P0) {
  [[clang::musttail]] return F11(1000);
}
```
```bash
clang -S -emit-llvm --target=(arm/x86/any 32 bit target) -o - 
```
Gives you (see #104770 for more info about weird no predecessors block):
```llvm
; Function Attrs: mustprogress noinline optnone
define dso_local arm_aapcscc void @_Z3F12s(ptr dead_on_unwind noalias writable sret({ double, double }) align 8 %agg.result, i16 noundef signext %P0) #0 {
entry:
  %P0.addr = alloca i16, align 2
  %tmp = alloca { double, double }, align 8
  store i16 %P0, ptr %P0.addr, align 2
  %0 = load i16, ptr %P0.addr, align 2
  musttail call arm_aapcscc void @_Z3F11s(ptr dead_on_unwind writable sret({ double, double }) align 8 %tmp, i16 noundef signext %0)
  ret void

1:                                                ; No predecessors!
...
  ret void
}
```
This is an invalid tail call. Similar behaviour can be observed when replacing `double _Complex` with `_BitInt(N>64)`.

### PR DESCRIPTION
**Pull request by kiran-isaac**
Fix #104794 

This fix adds test cases for `clang::musttail`, the final two of which would fail without my small change to CGCall.cpp

### CODE REVIEW

File: clang/lib/CodeGen/CGCall.cpp [hunk 1]
```diff
@@ -5085,7 +5085,7 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
   RawAddress SRetAlloca = RawAddress::invalid();
   llvm::Value *UnusedReturnSizePtr = nullptr;
   if (RetAI.isIndirect() || RetAI.isInAlloca() || RetAI.isCoerceAndExpand()) {
-    if (IsVirtualFunctionPointerThunk && RetAI.isIndirect()) {
+    if ((IsVirtualFunctionPointerThunk && RetAI.isIndirect()) || IsMustTail) {
```
Comment (efriedma-quic) [inline clang/lib/CodeGen/CGCall.cpp:5088]: Do any of the tests cover the case of `RetAI.isInAlloca()` or `RetAI.isCoerceAndExpand()`?

**Comment by kiran-isaac** [comment]
Accidentally pushed changes to this PR to main. Have reverted.

**Comment by kiran-isaac** [comment]
tagged wrong PR in revert commit, see #102896

File: clang/lib/CodeGen/CGCall.cpp (hunk 1 above)
Comment (efriedma-quic) [reply clang/lib/CodeGen/CGCall.cpp:5088]: Waiting for an answer here.

Comment (kiran-isaac) [reply clang/lib/CodeGen/CGCall.cpp:5088]: Apologies for slow replies on both of my PRs, our fuzzer and some of our downstream testing identified some issues and I had to verify they were unrelated. No none of the tests cover these cases. I am in the process of trying to add some that do.

**Comment by ostannard** [comment]
Continued in #109943.

**Pull request closed without merging**